emily/handler/src/api/handlers/chainstate.rs

```diff
@@ -50,13 +89,41 @@ pub fn get_chainstate(
         (status = 500, description = "Internal server error")
     )
 )]
-pub fn set_chainstate(
-    _request: SetChainstateRequestBody,
+pub async fn set_chainstate(
+    context: EmilyContext,
+    body: SetChainstateRequestBody,
 ) -> impl warp::reply::Reply {
-    let response = SetChainstateResponse {
-        ..Default::default()
-    };
-    with_status(json(&response), StatusCode::CREATED)
+    // Internal handler so `?` can be used correctly while still returning a reply.
+    async fn handler(
+        context: EmilyContext,
+        body: SetChainstateRequestBody,
+    ) -> Result<impl warp::reply::Reply, Error> {
+        // Create new chainstate table entry.
+        let new_chainstate_entry: ChainstateEntry = ChainstateEntry {
+            key: ChainstateEntryKey {
+                height: body.stacks_block_height,
+                hash: body.stacks_block_hash,
+            },
+        };
+
+        // TODO(TBD): Tactfully handle state inconsistencies and race conditions across
+        // multiple requests.
+        //
+        // - chain tip is ahead.
+        // - chain tip is far behind.
+        // - two chain tips are proposed at the same time for conflicting hashes
+        // - etc.
```

Comment:
Yes, I'm sorry to respond with this but this was part of what I had originally designed and while I can't remember 100% of why I remember strongly believing that it was exhaustive and necessary, and I believe I wrote it in the LLD. In this moment I need to lean on assuming my past assessment was correct and not redesign this.

That said, the overall idea of this API is to only allow a single path of reorganization; the put chainstate operation. If the chain tip is in the past and is a new block hash then the rest of the table will be updated, if the new tip is one above and new then continue on as you would otherwise, etc. And all the put and post updates from the other endpoints either append to the history or determine that the history inconsistent (I think) and do nothing.

Right now this is about the database accessors, I'll make another ticket to consider these options and map them back to the LLD.

Reply:
I apologize, this is a good line of questioning and suggesting, I'm not trying to shut this down entirely, only in this PR. I'll make space to discuss this explicitly later once the original design is implemented

Reply:
> I apologize, this is a good line of questioning and suggesting, I'm not trying to shut this down entirely, only in this PR. I'll make space to discuss this explicitly later once the original design is implemented

I didn't get the impression that you were shutting it down at all. And I think I have a better understanding of how you intend for this all to be used, so I'm glad that I asked.